### ISSUE TITLE
Read simulation status output into simulations database

### ISSUE THREAD

**Issue by jonrkarr** [reproducible-biomedical-modeling/CRBM-Viz#1851]
The simulator containers built with BioSimulators-utils (AMICI, COPASI, GillesPy2, PySCeS and soon BoolNet, BioNetGen, CBMPy, COBRApy) will save status information when they are next rebuilt.

These containers will also now produce reports (and eventually plots) progressively. Meaning, 
- Reports will be initialized once the first task completes. 
- Once at least one data set of a report has been generated, the HDF5 and CSV outputs will contain the data sets that have been generated
- All data sets which haven't been generated yet will have the value `NaN`
- Reports for which no data sets have been generated will have no columns
- Plots will behave similarly.
- Note: For now, reports are initialized with rows, but no other dimensions. The shape of the other dimensions is not realized until the first data has been produced. Its implemented this way because each KISAO term can produce a different shape of output (i.e., columns for non-spatial algorithms or multiple dimensions for spatial algorithms). We need the individual simulation tools to tell us this information about the other dimensions. The simulations are not yet passing such information to the initialization of the reports.

## Location
`{ out-directory }/status.yml`

### Rationale
I felt that status information should be saved to a separate file rather than entangled with reports (and HDF5) as we discussed for a few reasons
- Not all COMBINE archives generate reports. In fact, archives aren't required to have any outputs. Thus, we need to separate status and reports a bit
- It seemed to me that status files should be light-weight
- It seemed to me that status should either use a log-like format or something that can easily accommodate a straight-forward data structure. HDF5 doesn't work well for either.
- I chose YAML because its easily human-readable

### Alternatives
Obviously, YAML could easily be changed to JSON, which is slightly easier for use with TypeScript. But, I assume there's a YAML TypeScript package we can use to read these files.

### Data model
See https://github.com/biosimulators/Biosimulators_utils/blob/dev/biosimulators_utils/exec_status/data_model.py

### Example (in JSON format)
The potential status values are `QUEUED`, `RUNNING`, `SUCCEEDED`, `SKIPPED`, and `FAILED`.

```json
{
    "status": "SUCCEEDED",
    "tasks": {
        "task_1_ss": {"status": "SUCCEEDED"},
        "task_2_time_course": {"status": "SUCCEEDED"},
    },
    "outputs": {
        "report_1": {
            "status": "SUCCEEDED",
            "dataSets": {
                "dataset_1": "SUCCEEDED",
                "dataset_2": "SUCCEEDED",
            },
        },
        "report_2": {
            "status": "SUCCEEDED",
            "dataSets": {
                "dataset_3": "SUCCEEDED",
                "dataset_4": "SUCCEEDED",
            },
        },
        "report_3": {
            "status": "SUCCEEDED",
            "dataSets": {
                "dataset_5": "SUCCEEDED",
            },
        },
        "report_4": {
            "status": "SUCCEEDED",
            "dataSets": {
                "dataset_6": "SUCCEEDED",
            },
        },
    },
}
```

## Export schedule
The simulator containers built with BioSimulators-utils will output status to `{ out-directory }/status.yml` at the following times
- At the initiation of the execution of the COMBINE archive
  - The status of the archive will be set to `RUNNING`
  - The status of each SED document, task, and report will be set to `QUEUED`
  - The status of any task or plot which the simulator won't be able to execute (e.g., repeated tasks, plots) will be set to `SKIPPED`
- Before the execution of each SED-ML document
   - The document's status will be updated to `RUNNING`
 - After the execution of each SED-ML document
    - The document's status will be updated to `SUCCEEDED`
    - By this point, the status of all supported tasks and outputs in the document will have been set to `SUCCEEDED`
- Before the execution of each SED task
   - The task's status will be updated to `RUNNING`
 - After the execution of each SED task
    - The document's status will be updated to `SUCCEEDED    - `
    - The status of any reports and plots which now can be partially generated (i.e. some datasets or curves available) will be updated to `Running`
    - The status of any reports and plots which now can be fully generated (i.e., all data has been simulated) will be updated to `SUCCEEDED`

**Comment by bilalshaikh42** [comment]
Parsing yaml won't be a problem. Same steps json, but a different package. Simply do YAML.parse() rather than JSON.parse(). [yaml package](https://www.npmjs.com/package/yaml)

**Comment by bilalshaikh42** [comment]
1. Is the archive status ever set to `QUEUED`? Since we are having only one archive running at a time, I worry that a top level status will cause confusion with the status of a simulation run job. 
2. Can archives have a status of `SUCCEEDED` if a task or report was skipped due to not being supported? Can reports be successful independent of tasks that they depend on? 
3. Does  a simulation run that produces a status.yml with a archive status of `FAILED `result in a status of `FAILED` for the simulation-run? Or would this be successful, since the image worked as it should? I suppose this is the same issue as 1

**Comment by bilalshaikh42** [comment]
For updating the status to the database, is this also intended to be done through some sort of polling mechanism? This would be possible to implement but lead to some of the same issues that we are having currently with the NFS connection and SSH connection. It would also be another task that in kept in the dispatch-service as a stateful component. (#1711 #1714 #1712)

We could create some sort of a sidecar process in the simulator that can check for the presence of a token file. If this token is present, it can then use this token to submit the updates to the API directly.  The dispatch service would then ensure that this token is inserted for the jobs that are submitted.

**Comment by jonrkarr** [comment]
## 1
- The archive status is only briefly set to `QUEUED` 
- The archive status will be similar to the SLURM job status.
- I included the archive status for symmetry, and as a summary of the status of the entire job. I thought the summary would be useful when containers are run outside our SLURM environment, not necessarily in  the context of a HPC system which a job manager.

## 2
- Yes, an archive can have status `SUCCEEDED` even if task or report was `SKIPPED`. If helpful, we could introduce another status such as `SUCCEEDED_WITH_SKIPS`.
- If a task fails, then all reports that depend on the task will fail. This is also something we could change. Currently, jobs will exit as soon as the first task failure is encountered. In this case, the status report (`status.yml`) file is never updated with the status `FAILED`. 

  I could change this to the following. I haven't done this because the SLURM job management system is already collecting errors.
    1. run all tasks, catching all exceptions. 
    2. keep track of the exceptions of the tasks. 
    3. Generate a parts of reports and plots that can be generated (the parts that depend on tasks that did succeeded). 
    4. Save all of these complete and partial reports. 
    5. Throw an error if any task failed, in which case the whole process would exit with status != `0`.
    6. (SLURM would then recognize this non-zero exit status as a failed job.)

## 3
As I mentioned above, a task failure will cause the whole process to exit. In this case, `status.yml` file will never be updated with the status `FAILED`. 

## Updating the database
One possibility is for the simulation system to watch for changes to `status.yml`. 

Another possibility is to create an API endpoint which the simulator containers call when they update the status of their jobs. As you suggested, the containers could be given a token that they use to communicate which job they're reporting status for. This would be easy to incorporate into BioSimulators utils. In this case, they can post the status data in JSON format. They could even use a `PATCH` endpoint to only push the updated status.

**Comment by jonrkarr** [comment]
This isn't critical to implement now. I think it would be more impactful for us to focus on making the simulation system more reliable and then focusing our attention on the biosimulations.org sharing/reuse platform.

**Comment by jonrkarr** [comment]
A few days ago, I created biosimulators/Biosimulators_utils#15 to keep track of the simulator side of this issue.

Calling an API endpoint to post the status of a simulation is ideal. The only reason I was considering watching for changes to a YAML file is that we should be mindful of making it as easy as possible to create containers of simulation tools.

**Comment by jonrkarr** [comment]
Potential packages for rendering stdout/err
- ansi_up
- shell-color

**Comment by bilalshaikh42** [comment]
Done
